packages/ra-ui-materialui/src/button/BulkUpdateWithConfirmButton.tsx

```diff
@@ -0,0 +1,196 @@
+import * as React from 'react';
+import { FC, Fragment, useState, ReactElement } from 'react';
+import PropTypes from 'prop-types';
+import ActionUpdate from '@material-ui/icons/Update';
+import { fade } from '@material-ui/core/styles/colorManipulator';
+import inflection from 'inflection';
+import { makeStyles } from '@material-ui/core/styles';
+import {
+    useTranslate,
+    useUpdateMany,
+    useRefresh,
+    useNotify,
+    useUnselectAll,
+    CRUD_UPDATE_MANY,
+    useResourceContext,
+    MutationMode,
+} from 'ra-core';
+
+import Confirm from '../layout/Confirm';
+import Button, { ButtonProps } from './Button';
+import { BulkActionProps } from '../types';
+
+const useStyles = makeStyles(
+    theme => ({
+        updateButton: {
+            color: theme.palette.error.main,
+            '&:hover': {
+                backgroundColor: fade(theme.palette.error.main, 0.12),
+                // Reset on mouse devices
+                '@media (hover: none)': {
+                    backgroundColor: 'transparent',
+                },
+            },
+        },
+    }),
+    { name: 'RaBulkUpdateWithConfirmButton' }
+);
+
+const defaultIcon = <ActionUpdate />;
+
+const BulkUpdateWithConfirmButton: FC<BulkUpdateWithConfirmButtonProps> = props => {
+    const {
+        basePath,
+        mutationMode,
+        classes: classesOverride,
+        confirmTitle = 'ra.message.bulk_update_title',
+        confirmContent = 'ra.message.bulk_update_content',
+        data,
+        icon = defaultIcon,
+        label,
+        onClick,
+        selectedIds,
+        onSuccess = () => {
+            refresh();
```

Comment:
You're using refresh() before it's defined, same for other side effects. This may be valid but it's disturbing for the reader.